pipeline/run_pipeline.py

```diff
@@ -227,41 +242,43 @@ def plotting_refusal_with_activations(
 
 
 def computing_dot_product(
-    data_list,
-    refusal_direction,
-    average_dot_products,
-    resultant_dot_products,
-    activations_dir,
-    layers,
-    harmful_train,
-):
-    data_list = [normalize_vector(data) for data in data_list]
-
+    model_base,
+    layer_wise_activations: List[torch.Tensor],
+    refusal_direction: torch.Tensor,
+    average_dot_products: List[torch.Tensor],
+    resultant_dot_products: List[torch.Tensor],
+    activations_dir: str, 
+    layers: int, 
+    harmful_train: List[torch.Tensor]
+) -> None:
+    layer_wise_activations = [normalize_vector(data,model_base) for data in layer_wise_activations]
+
+    #layer_wise_activations is list, converting it to tensors for matching shape
+    layer_wise_activations_tensor = torch.stack(layer_wise_activations)
+    assert layer_wise_activations_tensor.shape == (len(harmful_train), model_base.model.config.num_hidden_layers, model_base.model.config.hidden_size)
+        
+    device=refusal_direction.device
     """Calculate and save dot products, then generate and save a DataFrame with results."""
     for i in range(layers):
         # Calculate average dot products
-        dot_products = sum(
-            np.dot(data_list[j][i].cpu().numpy(), refusal_direction.cpu().numpy())
-            for j in range(len(harmful_train))
-        ) / len(harmful_train)
-        average_dot_products.append(dot_products)
+        dot_products = torch.sum(torch.stack([torch.dot(layer_wise_activations[j][i].to(device), refusal_direction) for j in range(len(harmful_train))])) / len(harmful_train)
```

Comment:
```suggestion
        dot_products = (layer_wise_activations_tensor[:, i] @ refusal_direction).mean()
```
Stacking is very memory-inefficient (it creates new copies of your data in memory, which is very expensive). Use it in rare occasions only.